src/main/java/com/hotels/mutantswarm/MutantSwarmExtension.java

```diff
@@ -0,0 +1,211 @@
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+import org.junit.jupiter.api.extension.TestWatcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Resources;
+import com.klarna.hiverunner.HiveRunnerExtension;
+import com.klarna.hiverunner.HiveShellContainer;
+import com.klarna.hiverunner.annotations.HiveSQL;
+import com.klarna.hiverunner.builder.HiveRunnerScript;
+import com.klarna.hiverunner.builder.Script;
+import com.klarna.hiverunner.config.HiveRunnerConfig;
+import com.klarna.reflection.ReflectionUtils;
+
+import com.hotels.mutantswarm.MutantSwarmCore.ExecutionContext;
+import com.hotels.mutantswarm.exec.MutantState;
+import com.hotels.mutantswarm.exec.MutatedSourceFactory.MutatedSource;
+import com.hotels.mutantswarm.exec.SwarmResults;
+import com.hotels.mutantswarm.exec.SwarmResults.SwarmResultsBuilder;
+import com.hotels.mutantswarm.model.MutantSwarmScript;
+import com.hotels.mutantswarm.plan.Mutant;
+import com.hotels.mutantswarm.plan.Swarm;
+import com.hotels.mutantswarm.report.ReportGenerator;
+
+public class MutantSwarmExtension extends HiveRunnerExtension implements AfterAllCallback, TestWatcher,
+    TestTemplateInvocationContextProvider, TestInstancePostProcessor, AfterEachCallback {
+
+  private static final Logger log = LoggerFactory.getLogger(MutantSwarmExtension.class);
+  private static AtomicReference<ExecutionContext> contextRef = new AtomicReference<>();
+
+  public HiveShellContainer container;
+  private HiveRunnerConfig HiveRunnerConfig = new HiveRunnerConfig();
+  private int testNumber = -1;
+  private List<Mutant> mutants = new ArrayList<Mutant>();
+  private List<MutatedSource> mutatedSources = new ArrayList<MutatedSource>();
+  private ExecutionContext resultContext = contextRef.get();
+  private boolean firstTestPassed = true;
+  private MutantSwarmCore core = new MutantSwarmCore();
+
+  public MutantSwarmExtension() {}
+
+  @Override
+  public boolean supportsTestTemplate(ExtensionContext context) {
+    if (!context.getTestMethod().isPresent()) {
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public Stream<TestTemplateInvocationContext> provideTestTemplateInvocationContexts(ExtensionContext context) {
+    testNumber = -1;
+    firstTestPassed = true;
+
+    setFirstScripts(context);
+
+    if (contextRef.get() == null) {
+      Swarm swarm = core.generateSwarm(scriptsUnderTest, HiveRunnerConfig.getCommandShellEmulator());
+      SwarmResultsBuilder swarmResultBuilder = new SwarmResultsBuilder(swarm,
+          context.getRequiredTestClass().toString());
+      contextRef.compareAndSet(null, new ExecutionContext(swarm, swarmResultBuilder));
+    }
+
+    ExecutionContext executionContext = contextRef.get();
+    mutants = executionContext.swarm.getMutants();
+    for (Mutant mutant : executionContext.swarm.getMutants()) {
```

Comment:
Why not use the mutants variable since its defined above?
`for (Mutant mutant : mutants) { ` ?